src/mapdata.h

```diff
@@ -246,16 +248,28 @@ struct map_data_common_t {
 * Short for terrain type. This struct defines all of the metadata for a given terrain id (an enum below).
 */
 struct ter_t : map_data_common_t {
-    ter_id loadid;     // This is akin to the old ter_id, however it is set at runtime.
+    ter_str_id id;    // The terrain's ID. Must be set, must be unique.
+    ter_str_id open;  // Open action: transform into terrain with matching id
+    ter_str_id close; // Close action: transform into terrain with matching id
+
     std::string trap_id_str;     // String storing the id string of the trap.
     std::string harvestable;     // What will be harvested from this terrain?
-    std::string transforms_into; // Transform into what terrain?
-    std::string roof;            // What will be the floor above this terrain?
+    ter_str_id transforms_into; // Transform into what terrain?
+    ter_str_id roof;            // What will be the floor above this terrain
 
     trap_id trap; // The id of the trap located at this terrain. Limit one trap per tile currently.
 
     int harvest_season; // When will this terrain get harvested?
-    int bloom_season;   // When does this terrain bloom?
+
+    ter_t() :
+        open( NULL_ID ),
+        close( NULL_ID ),
+        transforms_into( NULL_ID ),
+        roof( NULL_ID ),
+        trap( tr_null ),
```

Comment:
Because of the implicit caching in the string_id, you can remove the `trap` member. One can use the `trap_id_str` member directly. Maybe rename it it `trap` so existing code may not need to be changed.